### PR TITLE
[SPARK-35564][SQL] Improve subexpression elimination

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?

This PR proposes a new way to do subexpression elimination in `EquivalentExpressions`. The main change of the PR is that `ExpressionStats` stores the expected evaluation count of subexpressions split into `evalCount` that records sure evaluations and `conditionalEvalCount` that records expected conditional evaluations.

Please note that the expected conditional evaluation count is not the same as how many times a subexpression appears in `EquivalentExpressions` conditionally (conditional use count). The expected conditional evaluation count better describes how likely a conditional subexpression is evaluated and so we can use it easier to define when we should consider a subexpression for elimination (e.g. some kind of threshold is reached).
The idea behind using expected conditional evaluation count is that if we consider 2 cases where a `c` non-leaf subexpression:
- appears only in one branch of a simple 2-branched `If` expression
- or it appears only in the last branch of a 10-branched `CaseWhen` expression,

then the conditional use count is 1 in both cases. But the expected conditional evaluation counts are different. Very likely `c` will be evaluated more in the first case if we consider random input data. Since we don't know the exact probabilities of the branches, for the sake of simplicity all branchings are modelled with 0.5 / 0.5 probabilities in this PR.

Please find a related conversation here about the default 0 value of the config: https://github.com/apache/spark/pull/32987#issuecomment-1258168996

Here are a few example expressions and the `ExpressionStats` (sure + expected conditional evaluation counts) of a non-leaf `c` subexpression from the equivalence maps built from the expressions:

| Expression | `ExpressionStats` of `c` |
|---|---|
|`c` | `c -> (1 + 0.0)` |
|`c + c` | `c -> (2 + 0.0)` |
|`If(_, c, _)` | `c -> (0 + 0.5)` |
|`If(_, c + c, _)` | `c -> (0 + 1.0)` |
|`If(_, c, c)` | `c -> (1 + 0.0)` |
|`If(c, c, _)` | `c -> (1 + 0.5)` |
|`If(c, c, c)` | `c -> (2 + 0.0)` |

This PR:
- Fixes the issue of subexpressions that are surely evaluated only once but there is a certain probability that they are evaluated more. These subexpressions are now considered for elimination based on the newly introduced `spark.sql.subexpressionElimination.minExpectedConditionalEvaluationCount` config.
- Fixes the issue of branch groups in `CaseWhen` and `Coalesce` expressions. Branch groups were used for calculating common subexpressions in conditional branches based on the idea that subexpressions that appear in all elements of a group are surely evaluated once. If we take the `CaseWhen(w1, t1, w2, t2, w3, t3, e)` example then the previously defined (`t1`, `t2`,`t3`, `e`) group made sense, but for some reason the (`w1`, `w2`, `w3`) group was also defined, which didn't make sense because `w1` was also considered always evaluated. Also, some other groups that would have made sense (`t1`, `w2`) and (`t1`, `t2`, `w3`) were not defined. This PR completely removes branch groups from `ConditionalExpression` and uses a new way to calculate surely evaluated subexpressions.

### Why are the changes needed?
Improve subexpression elimination.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing and new UTs (including the ones from @Kimahriman's PR: https://github.com/apache/spark/pull/32987).

### CODE REVIEW

**Comment by peter-toth** [comment]
This PR is still WIP because I want to add more tests, but @Kimahriman, @cloud-fan, @rednaxelafx, @ulysses-you, @viirya, @wankunde you might be interrested...

**Comment by Kimahriman** [comment]
This hurts my brain thinking about probabilistic conditional evaluations, and I feel like the subexpression elimination logic is already overly complicated. If I wanted to just create subexpressions for anything that is definitely executed once and maybe executed one other time (regardless of how nested inside a CaseWhen or Coalesce operation), what do I even set the new setting to?

**Comment by peter-toth** [comment]
> This hurts my brain thinking about probabilistic conditional evaluations, and I feel like the subexpression elimination logic is already overly complicated. If I wanted to just create subexpressions for anything that is definitely executed once and maybe executed one other time (regardless of how nested inside a CaseWhen or Coalesce operation), what do I even set the new setting to?

Got it, thanks for your feedback. If `conditionalEvalCount` is considered is an overkill then I can revert it `conditionalUseCount` or a simple boolean flag. BTW, I think with this PR the `ExpressionStats` calculation logic becomes much simpler than it was before (especially if we revert to `conditionalEvalCount` or a boolean flag), the `getCommonSubexpressions()` method is what became a bit more complicated. 

Currently the new config is used as `conditionalEvalCount >= <config value>` so you could use a very small value to behave the same way as `conditionalUseCount > 0`. Or we can change the config semantics to `>` and use 0 config value for the same effect.

**Comment by Kimahriman** [comment]
If you just have `conditionalUseCount` than that's basically what I have in my PR I think (except true/false currently in mine instead of number of conditional uses). Removing the "branch groups" definitely simplifies things though if we're just willing to make those be conditional uses instead, that was the hardest part of getting the conditional usage working.

**Comment by peter-toth** [comment]
> Removing the "branch groups" definitely simplifies things though if we're just willing to make those be conditional uses instead

No, I mean although I removed `branchGroups` (and `alwaysEvaluatedInputs`) from `ConditionalExpression`, this PR maintains the count of sure evaluations correctly in `ExpressionStats.evalCount` as before. See `If(_, c, c)` => `c ->(1 + 0.0)` example in the description. And actually it is better than before in case of `CaseWhen`...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 1]
```diff
@@ -32,209 +32,325 @@ import org.apache.spark.util.Utils
...
+          // between `T1` and `W2`. It is an intersection in terms of surely evaluated
+          // subexpressions (`evalCount`) between `T1` and `W2` but it is also kind of an union
+          // between conditionally evaluated subexpressions (`conditionalEvalCount`). See the
+          // details in `intersectMapWithMap()`.
+          // So the local map can be calculated as `W1 | (T1 & W2)` so far, where `|` and `&` mean
+          // the "union" and "intersection" of equivalence maps.
+          // But we can continue the previous logic further because if `w2` is evaluated, then based
+          // on the result of `w2` either `t2` or `w3` is also evaluated.
+          // So eventually the local equivalence map can be calculated as
+          // `W1 | (T1 & W2 | T2 & (W3 | T3 & ... & (Wn | Tn & E)))`.
```
Comment (Kimahriman) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:111]: Is there a missing parenthesis here? I'm trying to understand the order of operations once you get to `T2`.
Is it `W1 | T1 & (W2 | T2 & (W3 | ...`
assuming normal higher precedence of `&`

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala]: \+ I've changed the new config in: https://github.com/apache/spark/pull/41677/commits/38a0996aa1dbf2471e6a5d4fa4dc031995b9beb0 and from now the default 0 value has the same effect as `spark.sql.subexpressionElimination.conditionals.enabled` in your PR.

**Comment by peter-toth** [comment]
The failure in `[Run / Linters, licenses, dependencies and documentation generation]` seems unrelated.

**Comment by peter-toth** [comment]
I've added a few more test cases and this PR is now ready for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 2]
```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
...
  */
 class EquivalentExpressions(
-    skipForShortcutEnable: Boolean = SQLConf.get.subexpressionEliminationSkipForShotcutExpr) {
+    skipForShortcutEnable: Boolean = SQLConf.get.subexpressionEliminationSkipForShotcutExpr,
+    minConditionalCount: Option[Double] =
+      Some(SQLConf.get.subexpressionEliminationMinExpectedConditionalEvaluationCount)
+        .filter(_ >= 0d),
+    allowLeafExpressions: Boolean = false) {
+
+  // The subexpressions are stored by height to speed up certain calculations.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:46]: sorted by height?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 3]
```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
  * This class is used to compute equality of (sub)expression trees. Expressions can be added
  * to this class and they subsequently query for expression equality. Expression trees are
  * considered equal if for the same input(s), the same result is produced.
+ *
+ * Please note that `EquivalentExpressions` is mainly used in subexpression elimination where common
+ * non-leaf expression subtrees are calculated, but there there is one special use case in
+ * `PhysicalAggregation` where `EquivalentExpressions` is used as a mutable set of non-deterministic
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:36]: hmm, how can `EquivalentExpressions` handle non-deterministic expressions?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 4]
```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
...
+  private var inflated: Boolean = true
 
   /**
-   * Adds each expression to this data structure, grouping them with existing equivalent
-   * expressions. Non-recursive.
-   * Returns true if there was already a matching expression.
+   * Adds each expression to this data structure and returns true if there was already a matching
+   * expression.
    */
   def addExpr(expr: Expression): Boolean = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:65]: it looks like the only difference between this and `addExprTree` is, `addExprTree` allows non-deterministic expression. Shall we name these two methods better?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 3 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:36]: Yeah, this is not right. I should have written `mutable set of deterministic expressions`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 2 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:46]: `maps` is an array buffer and each element stores a map that contains expressions with certain height. The `i`th element contains a map of expressions with height `i+1`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 5]
```diff
@@ -243,28 +414,43 @@ class EquivalentExpressions(
...
 
-  override def hashCode: Int = Objects.hash(e.semanticHash(): Integer, height: Integer)
+  override def hashCode: Int = Objects.hash(expr.semanticHash(): Integer, height: Integer)
 }
 
 /**
- * A wrapper in place of using Seq[Expression] to record a group of equivalent expressions.
+ * This class stores the expected evaluation count of expressions split into `evalCount` +
+ * `realEvalCount` that records sure evaluations and `condEvalCount` + `realCondEvalCount` that
+ * records conditional evaluations. The `real...` fields are filled up during `inflate()`.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:438]: I don't quite get the meaning of the `real...` fields by reading the comment here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 6]
```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
...
   }
 
-  private def supportedExpression(e: Expression) = {
-    !e.exists {
-      // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
-      // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
-      case _: LambdaVariable => true
+  private def inflateExprState(exprStats: ExpressionStats): Unit = {
+    val expr = exprStats.expr
+    if (!expr.isInstanceOf[LeafExpression] || allowLeafExpressions) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:140]: what will go wrong if we always allow leaf expressions?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 5 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:438]: I agree, `direct` and `transitive` are better prefixes.

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:438]: Btw `transitive` ( currently `real`) is actually a sum of the direct additions and the transitive additions from parent expressions.
If we `addExprTree(a + b)` and `addExpr((a + b) + 1)` then the direct of `a + b` is 1 and the transitive of `a + b` is 2.
 
We didn't need the `transitive` (`real`) fields if did recurse during `addExprTree()`.
But you know, the previous version of `EquivalentExpressions` used `useCount`. And with `useCount` when same or overlapping expressions were added to the data structure the second `addExprTree()` didn't fully recurse, but it stopped when the first common subexpression was found. Now with the new structure, we just record the `direct` additions during `addExprTree()` and fill the `transitives` during `inflate()` to be par on with the old version.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 7]
```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
...
+    zippedMaps.foreach { case (map, otherMap) =>
+      map.foreach { case (key, value) =>
+        otherMap.remove(key) match {
+          case Some(otherValue) =>
+            val (min, max) = if (value.realEvalCount < otherValue.realEvalCount) {
+              (value.realEvalCount, otherValue.realEvalCount)
+            } else {
+              (otherValue.realEvalCount, value.realEvalCount)
+            }
+            value.realCondEvalCount += otherValue.realCondEvalCount + max - min
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:270]: shouldn't this be `value.realCondEvalCount = (value.realCondEvalCount + otherValue.realCondEvalCount) / 2`?

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:270]: `value.realCondEvalCount = (value.realCondEvalCount + otherValue.realCondEvalCount + max - min) / 2` is the full calculation, but the `value.realCondEvalCount /= 2` extracted a bit below.

The `max - min` / 2 is also need. E.g. if we have `If(_, a + b, (a + b) + (a + b))` then during the intersect of the `then` and `else` branches we have `a + b -> 1 + 0` in `then` and `a + b -> 2 + 0` in `else`. The result should be (`a + b -> 1 + 0.5)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala (hunk 6 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:135]: We don't need to store leaf expressions when we use `EquivalentExpressions` for CSE as leafs don't make sense to evaluate in advance. `addExprTree()` didn't recurse to leafs in the previous version but now that `addExprTree()` doesn't recurse, we need this flag.

**Comment by peter-toth** [comment]
@cloud-fan, @Kimahriman please let me know if we should take this PR further, otherwise I let this PR closed by the automation.

**Pull request closed without merging**